### DISCUSSION TITLE
Погода от Яндекс

### DISCUSSION THREAD

**Discussion by shkolin1973** [vvip-68/GyverPanelWiFi#303] [Задавайте вопросы]
Здравствуйте. 27 июня 2024 года перестал работать запрос погоды от Яндекс (65 регион Новосибирск) В логе ошибка подключения. Не работает на двух устройствах.

**Comment by vvip-68** [comment]
Яндекс зачем-то поменял формат ответа на запрос погоды. По ошибке или сознательно - не знаю. Навскидку я вижу лишний перевод строки перед телом JSON ответа. Поменялись ли какие другие поля - пока детально не смотрел.
Тут  следующие варианты:
1. Самое быстрое - настроить и переключиться на получение погоды с OpenWeatherMap
2. Подождать, может яндекс вернет как было...
3. Попытаться в коде разбора ответа с погодой удалить этот лишний перевод строки и если не поможет - смотреть что еще пошло не так, есть ли в ответе нужные поля, не переименованы ли они и так далее.
4. Ждать пока сделать все это руки дойдут у меня. Времени пока особо нет. Могу достаточно быстро посмотреть что не так в версии 1.13. А вот 1.14 - в полуразобранном состоянии - недоделанные и брошенные за неимением времени переделки. Нужно или доделывать их или откатывать назад...
--
Отправлено из Mail.ru для Android пятница, 28 июня 2024г., 09:44 +07:00 от shkolin1973  ***@***.*** :

**Comment by shkolin1973** [comment]
Спасибо за ответ! Значит всё же "лыжи не едут" :-)) У меня устройства на версии 1,13. Мне нравится. Спасибо вам за труды! Будет время, исправьте пожалуйста.

**Reply by vvip-68** [reply]
Посмотрел ответ от сервиса. Яндекс не формат поменял. Он перестал отвечать на короткий запрос погоды, возвращая код 302 - Moved temporarily. :( Видимо слишком много устройств с этим принципом запроса стали обращаться к серверу и они отключили обработку таких запросов.

На запрос из браузера ответ нормальный. Но браузер шлет на сервер больше заголовков в запросе.
Надо в программе для контроллера что-то дополнительное отправлять на сервер, чтобы он думал, что это не от контроллеров запрос идет, а, например, из браузера.. Что - пока не понял... Надо экспериментировать. Если время и желание будет. Ну или пока кто-нибудь не раскопает как правильно формировать запрос и не поделится своими знаниями.

Пока же придется переключиться на получение погоды с OpenWeatherMap.

**Reply by vvip-68** [reply]
Похоже, яндекс перенастроил свой сервер и если он раньше отвечал на запрос по нешифрованному каналу http:// теперь вместо ответа перенаправляет на шифрованный - говорит, что обязательно нужно использовать https:// с шифрованием
А это, в общем, морока. Я не умею :)
Тут надо сертификаты формировать / подключать и клиент использовать не WiFiClient, а WiFiClientSecure с указанием сертификата.
В общем, будем разбираться как время будет...

**Reply by shkolin1973** [reply]
OpenWeatherMap -работал нормально, теперь начал глючить. И такое ощущение это подвешивает контроллер. Он периодически стал перегружаться.

**Reply by vvip-68** [reply]
Ну, надо смотреть хотя бы что в логе пишет.
Визуально у меня два устройства в зоне видимости работают беспроблемно с OpenWeatherMap...

**Reply by shkolin1973** [reply]
Глючит у меня устройство на базе Wemos. У меня уже возникли подозрения не загнул ли он "ласты".
Достал устройство на базе ESP32, пол года лежал в тумбочке без включения.
Сначала всё было хорошо, тыкал в приложении кнопку "Установить" в поле выбора сервера погоды и погода успешно обновлялась (смотрел в мониторе), а потом пошли ошибки и и дут до сих пор. Но он не зависает и не перегружается. А не связано ли это с количеством запросов в единицу времени от конкретного устройства? Вот лог ESP32:

08:56:05.976 -> Запрос текущей погоды
08:56:06.352 -> Погода получена: OpenWeatherMap
08:56:06.352 -> Город: Новосибирск
08:56:06.352 -> Сейчас: ясно, +20ºC
08:56:06.352 -> Код иконки: '01d'
08:56:06.352 -> Код погоды: 800
08:56:06.352 -> Сейчас: Светлое время суток
08:56:06.352 -> Рассвет: 05:18
08:56:06.352 -> Закат: 21:50
08:56:06.352 -> 
08:56:06.352 -> MQTT >> OK >> [ ] user/shkolin/panel4/wtr	 >> {"act":"WEATHER","region":1496747,"result":"OK","status":"ясно","temp":20,"night":false,"icon":"01d","town":"Новосибирск","server":"OpenWeatherMap","code":800,"sunrise":"05:18","sunset":"21:50"}
08:56:08.697 -> UDP << ip='192.168.1.3:2394'; cmd='$12 4 2;'
08:56:08.697 -> Получен UDP пакeт размером 8
08:56:08.697 -> UDP 192.168.1.3:2394 >> $18 UP:539|FM:200036|AL:0|W:16|H:16|DM:0|PS:1|PD:60|IT:0|RM:0|PW:1000|BR:19|WU:2|WT:15|WR:65|WS:1496747|WC:1|WN:0|WZ:1|SD:0|FS:1|EE:1;
08:56:08.744 -> 
08:56:08.744 -> Запрос текущей погоды
08:56:09.916 -> Ошибка сервера погоды: 
08:56:09.916 -> MQTT >> OK >> [ ] user/shkolin/panel4/wtr	 >> {"act":"WEATHER","region":1496747,"result":"ERROR","status":""}
08:56:13.715 -> Таймаут запроса погоды!
08:56:13.715 -> 
08:56:13.715 -> Запрос текущей погоды
08:56:14.935 -> Ошибка сервера погоды: 
08:56:14.935 -> MQTT >> OK >> [ ] user/shkolin/panel4/wtr	 >> {"act":"WEATHER","region":1496747,"result":"ERROR","status":""}
08:56:14.935 -> UDP << ip='192.168.1.3:2394'; cmd='$18 0;'
08:56:14.935 -> Получен UDP пакeт размером 6
08:56:14.935 -> Ответ на 192.168.1.3:2394 >> ack7;
08:56:18.710 -> Таймаут запроса погоды!
08:56:18.710 -> 
08:56:18.710 -> Запрос текущей погоды
08:56:19.836 -> Ошибка сервера погоды: 
08:56:19.836 -> MQTT >> OK >> [ ] user/shkolin/panel4/wtr	 >> {"act":"WEATHER","region":1496747,"result":"ERROR","status":""}
08:56:19.836 -> UDP << ip='192.168.1.3:2394'; cmd='$18 0;'
08:56:19.836 -> Получен UDP пакeт размером 6
08:56:19.883 -> Ответ на 192.168.1.3:2394 >> ack8;
08:56:23.717 -> Таймаут запроса погоды!
08:56:23.717 -> 
08:56:23.717 -> Запрос текущей погоды
08:56:24.878 -> Ошибка сервера погоды: 
08:56:24.878 -> MQTT >> OK >> [ ] user/shkolin/panel4/wtr	 >> {"act":"WEATHER","region":1496747,"result":"ERROR","status":""}
08:56:24.878 -> UDP << ip='192.168.1.3:2394'; cmd='$18 0;'
08:56:24.878 -> Получен UDP пакeт размером 6
08:56:24.878 -> Ответ на 192.168.1.3:2394 >> ack9;
08:56:28.723 -> Таймаут запроса погоды!
08:56:28.723 -> 
08:56:28.723 -> Запрос текущей погоды
08:56:29.895 -> Ошибка сервера погоды: 
08:56:29.895 -> MQTT >> OK >> [ ] user/shkolin/panel4/wtr	 >> {"act":"WEATHER","region":1496747,"result":"ERROR","status":""}
08:56:29.895 -> UDP << ip='192.168.1.3:2394'; cmd='$18 0;'
08:56:29.895 -> Получен UDP пакeт размером 6
08:56:29.895 -> Ответ на 192.168.1.3:2394 >> ack10;
08:56:33.733 -> Таймаут запроса погоды!
08:56:33.733 -> 
08:56:33.733 -> Запрос текущей погоды
08:56:34.905 -> Ошибка сервера погоды: 
08:56:34.905 -> MQTT >> OK >> [ ] user/shkolin/panel4/wtr	 >> {"act":"WEATHER","region":1496747,"result":"ERROR","status":""}
08:56:34.905 -> UDP << ip='192.168.1.3:2394'; cmd='$12 4 2;'
08:56:34.905 -> Получен UDP пакeт размером 8
08:56:34.905 -> UDP 192.168.1.3:2394 >> $18 UP:566|FM:195064|AL:0|W:16|H:16|DM:0|PS:1|PD:60|IT:0|RM:0|PW:1000|BR:19|WU:2|WT:15|WR:65|WS:1496747|WC:1|WN:0|WZ:1|SD:0|FS:1|EE:1;
08:56:34.905 -> 
08:56:34.905 -> Запрос текущей погоды
08:56:36.124 -> Ошибка сервера погоды: 
08:56:36.124 -> MQTT >> OK >> [ ] user/shkolin/panel4/wtr	 >> {"act":"WEATHER","region":1496747,"result":"ERROR","status":""}
08:56:36.124 -> UDP << ip='192.168.1.3:2394'; cmd='$12 4 1;'
08:56:36.124 -> Получен UDP пакeт размером 8
08:56:36.124 -> UDP 192.168.1.3:2394 >> $18 UP:567|FM:195056|AL:0|W:16|H:16|DM:0|PS:1|PD:60|IT:0|RM:0|PW:1000|BR:19|WU:1|WT:15|WR:65|WS:1496747|WC:1|WN:0|WZ:1|SD:0|FS:1|EE:1;
08:56:36.171 ->

**Reply by shkolin1973** [reply]
Снова заработало нормально. Сделал 50 принудительных запросов погоды - работает. Х.З. в чем дело, у меня больше нет предположений.

**Reply by vvip-68** [reply]
Простой гуглеж по строке 'Brownout detector was triggered' говорит, что проблемы с питанием устройства.
Может отсюда ноги растут?

**Reply by shkolin1973** [reply]
С питанием был рамс, подключил от порта компьютера, без матрицы работало, с матрицей начало глючить. Я удалил это сообщение, но вы похоже успели его прочитать. 
Но выяснил, что запрос не работает или работает синхронно на 2-х устройствах:
1. Wemos там я получил и вписал свой API
2. ESP32 там, ваш API я не менял
тут или с сайтом что-то или с моим интернетом.

**Reply by vvip-68** [reply]
Вот с ключом API, кстати, интересный вопрос. По умолчанию там стоит какой-то demo-ключ. Если его никто не меняет, ленясь создать свой, то тысячи запросов с одним и тем же ключом могут поступать ежесекундно на сервер. Полагаю, что сервер может просто блокировать такие запросы... Не в этом ли причина?

Я не могу сказать, что идет не так... У себя я таких проблем не замечаю...

**Reply by shkolin1973** [reply]
На одном устройстве я поменял API, там где указано в инструкции. Может нужно поменять где-то ещё?

**Reply by vvip-68** [reply]
Да нет, ключ в одном месте хранится...

**Reply by shkolin1973** [reply]
Не зависит от API (сформировал ещё один и записал в Wemos ). Либо оба работают, либо нет. И похоже эти запросы сильно нагружают контроллер. Особенно запросы на OpenWeatherMap. До такой степени, что Wemos очень сильно подвисает, не подключается к приложению, а когда приложение (с 25 раза подключается) с большим запозданием реагирует на команды. У ESP32 такой проблемы нет (не виснет).

**Reply by vvip-68** [reply]
Отличие ESP32 от Wemos - ESP32 двухядерный. По умолчанию система крутится на одном ядре, пользовательский скрипт - на другом. Запросы к серверу погоды - синхронные. То есть - отправили запрос - ждет ответа. Пока ждет ответа ничего другого не выполняется. Вроде так. Это касательно одноядерного Wemos. Если сервер долго не отвечает - беда-беда... Если совсем долго нет ответа - Watchdog считает что скрипт завис и перезагружает контроллер. Вероятно это и происходит. Другой вопрос - почему сервер не отвечает... Но тут уже дело не в скрипте матрицы...

**Reply by shkolin1973** [reply]
Да, похоже всё так и есть. Пока сделал период запроса 30мин, было 15мин. Железо действительно целое, я думал Wemos сдох. Почему сервис периодически не доступен - открыт. 
Скрипт вне подозрения он исправно работал пол года на Яндексе, и 3 недели на OpenWeatherMap, глюки начались несколько дней назад.

**Reply by vvip-68** [reply]
Вероятно пора закрывать проект. Что-то меняется, часть функционала перестает работать. Появляется много вопросов. А исправить я это не могу...

**Reply by shkolin1973** [reply]
Всё, решение есть! Проблема DNS провайдера, что с ними стало Х.З. Их в мой роутер прописывается два автоматом. Я поменял один на 8.8.8.8 и всё заработало!

**Reply by vvip-68** [reply]
Обновил сегодня проект. Поправил получение погоды с Яндекса.
Получение идет через сторонний прокси-сервер. Через него теперь, видимо, будут идти запросы от всех, кто пользуется прошивкой. Поэтому, чтобы Яндекс не заблокировал IP сервера - желательно ставить получение погоды не чаще раз в 30, а лучше - раз в 60 минут...

**Reply by shkolin1973** [reply]
Спасибо!

**Reply by 33tooth** [reply]
Здравствуйте, как обновить прошивку чтобы Яндекс погода заработала?

**Reply by vvip-68** [reply]
Скачать с репозитория последнюю версию и обновить обычным способом.

**Reply by 33tooth** [reply]
скачал последнюю версию, при попытке компиляции выводит ошибку:  fatal error: a_def_hard.h: No such file or directory

**Reply by vvip-68** [reply]
Значит что-то не так скачали, не туда распаковали и так далее.
Вот этот файл в репозитории - он есть:
![image](https://github.com/user-attachments/assets/d1d61845-0551-4bde-a6e9-8805cdf645b7)

**Comment by shkolin1973** [comment]
Обнаружилось, что не загружаются файлы карт индексов
 09:05:01.854 -> Поиск файлов карт индексов...
09:05:01.854 -> Нет файлов карт индексов
Заранее отвечаю на вопрос, да стоит разрешение матрицы 16х16 и файл индексов из проекта.
Файлы эффектов и слайды загружаются и работают как надо.
Если загрузить карту индексов в память МК, она подгружается нормально.
Мне это не нужно, но вот так...

**Reply by vvip-68** [reply]
Ну так скетч и написан - карта индексов может быть ТОЛЬКО в памяти микроконтроллера...
Если вам нужно как-то по другому - дописывайте необходимый вам функционал.

**Reply by shkolin1973** [reply]
Хорошо, значит всё правильно работает! Вроде стараюсь все комментарии читать, но тут только с Вашей подачи смог найти место где это написано.
Спасибо за Вашу работу! Проект очень нравится!
